### PR TITLE
Fix #86: Add API to restart one or more kafka broker(s)

### LINKED ISSUE #86: API to restart kafka brokers
Thinking forward about testing Kroxylicious edge cases - network disconnection.

As the author of a component like Kroxylicious, I want to be able to write integration tests that ensure my component correctly handles broker restarts correctly.  To this end I would value an API that permitted me to restart one or more brokers within a cluster.  It is important that the broker comes back on the same ports (as it would in production).  I have this requirement for both the in-VM and container use-case.

I'm imagine a API that looks like:

`#restart(int nodeId)`
or maybe 
`#restart(Set<Integer> nodeId)`

control would return to the caller once the  broker restart is finished and all the brokers see one and other.

I don't envisaged needing functionality to do things like roll the brokers of a cluster (waiting for ISRs etc).  If a test needs that, it could easily construct it itself.

I don't envisage needing the ability to restart any kafka broker that is solely in the kraft controller role.

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

Add API to restart one or more kafka broker(s).

### Additional Context

This is done so that users of the test framework may exercise code paths where the broker(s) of the cluster are restarted, thus ensuring the robustness of error handling code etc.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 1]
```diff
@@ -320,21 +321,27 @@ public synchronized void removeBroker(int nodeId) throws IllegalArgumentExceptio
         s.awaitShutdown();
     }
 
+    @Override
+    public void restartBrokers(Predicate<Integer> nodeIdPredicate, boolean abruptShutdown) {
+        var kafkaServersToRestart = servers.entrySet().stream().filter(e -> nodeIdPredicate.test(e.getKey()))
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:326]: Fixed.

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 2]
```diff
@@ -309,26 +317,112 @@ public synchronized void removeBroker(int nodeId) throws UnsupportedOperationExc
...
+                // seems to lead to podman unreliability (at least on the Mac).
+                kc.withStartupAttempts(1);
+                Unreliables.retryUntilSuccess(Long.valueOf(STARTUP_TIMEOUT.toMillis()).intValue(), TimeUnit.MILLISECONDS,
+                        () -> {
+                            try {
+                                kc.start();
+                            }
+                            catch (Exception e) {
+                                kc.stop();
+                                long backoff = 5000;
```
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:406]: A fixed 5s feels quite slow. I wonder how much extra complexity we'd incur to have exponential backoff?

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 3]
```diff
@@ -309,26 +317,112 @@ public synchronized void removeBroker(int nodeId) throws UnsupportedOperationExc
...
+
+        LOGGER.log(Level.DEBUG, "Restarting {0}/{1} broker(s)", kafkaContainersToRestart.size(), getNumOfBrokers());
+        kafkaContainersToRestart.forEach(kc -> {
+            var originalStartupAttempts = kc.getStartupAttempts();
+            try {
+                // We need to control the restart loop ourselves. Unfortunately testcontainers does not
+                // allow a delay to be introduced between startup attempts and start-ups in a tight loop
+                // seems to lead to podman unreliability (at least on the Mac).
+                kc.withStartupAttempts(1);
+                Unreliables.retryUntilSuccess(Long.valueOf(STARTUP_TIMEOUT.toMillis()).intValue(), TimeUnit.MILLISECONDS,
```
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:399]: I went to check the `Unreliables` API looking for backoff options and noticed the [repo](https://github.com/rnorth/duct-tape) has been archived. Do we think its good enough or should we look at moving to awaitility or something else?

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java [hunk 4]
```diff
@@ -112,6 +117,33 @@ public void kafkaClusterZookeeperRemoveBroker() throws Exception {
...
+                Arguments.of(1, true, true, (Predicate<Integer>) node -> true),
+                Arguments.of(2, true, true, (Predicate<Integer>) node -> node == 1),
+                Arguments.of(2, true, true, (Predicate<Integer>) node -> true),
+                Arguments.of(1, true, false, (Predicate<Integer>) node -> true),
+                Arguments.of(1, false, true, (Predicate<Integer>) node -> true));
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    public void restartBrokers(int brokersNum, boolean kraft, boolean abruptShutdown, Predicate<Integer> integerPredicate) throws Exception {
```
Comment (SamBarker) [inline impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:131]: The test doesn't actually verify that anything was restarted 😞 

The test would pass if `cluster.restartBrokers(integerPredicate, abruptShutdown);` was entirely a no-op.

Comment (franvila) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:131]: Totally agree with @SamBarker, we would need to check somehow that the brokers has been restarted or going down for a while. Is there any property that is changed when a broker is restarted (thinking about nodeId or something similar)?

File: impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java [hunk 5]
```diff
@@ -267,6 +268,7 @@ public ConfigHolder generateConfigForSpecificNode(KafkaEndpoints kafkaEndpoints,
             putConfig(server, "zookeeper.connect", kafkaEndpoints.getEndpointPair(Listener.CONTROLLER, 0).connectAddress());
             putConfig(server, "zookeeper.sasl.enabled", "false");
             putConfig(server, "zookeeper.connection.timeout.ms", Long.toString(60000));
+            putConfig(server, KafkaConfig.ZkSessionTimeoutMsProp(), Long.toString(6000));
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/common/KafkaClusterConfig.java:271]: Perhaps a separate PR updating all the other keys to use their respective property method would be best.

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java (hunk 4 above)
Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:131]: Thank you - that's a good spot.   It is not immediately obvious to me how to tell this from the kafka client.    AFAIK brokers don't have any kind of 'session/epoch' id available via the client.  Detecting the client's disconnection would be an imperfect proxy for a broker restart (but could suffer from false positives), but I don't actually see a clean way to do that either!

I'd rather not break the abstraction and start doing horrid things like searching for patterns in broker logs files.

Thinking hat time. Ideas welcome.

Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:131]: I suppose one way to approach the problem might be to have the API accept a callback. 

```java
cluster.restartBrokers(Predicate<Integer>, boolean, Callable<Void> brokersDown);
```

 The semantics would be that the `restartBrokers` invokes `brokersDown` once the broker(s) are down. Restart wouldn't continue until the callback returns.  The test could then use the callback to make an independent verifications. 

It might have utility in Kroxylicious's test suite too - I can imagine it making an observation of say a target cluster's state, verifying that it has transitioned to, say,  'OFFLINE'.

Comment (SamBarker) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:131]: I like the suggestion I'm not entirely convinced about needing to wait for the callback to complete before continuing.

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java (hunk 2 above)
Comment (SamBarker) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:406]: Yeah I hear you! I am all for doing something here to increase reliability of the tests.

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java (hunk 4 above)
Comment (franvila) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:131]: I like it too. It is an smart way to say "Hey I'm down and going back online". Thanks @k-wall!

File: api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java [hunk 6]
```diff
@@ -42,6 +43,22 @@ public interface KafkaCluster extends AutoCloseable {
...
+     * The caller may specify <code>abruptShutdown</code> true.  If this is the case, the kafka broker
+     * will be abruptly killed, rather than undergoing a graceful shutdown. This may be useful to test cases
+     * wishing to explore networking edge cases.  The implementation may ignore this parameter if the
+     * implementation is unable to support the requested style of shutdown. In this case, the implementation
+     * is free to use an alternative shutdown style instead.
+     * <br/>
+     * @param nodeIdPredicate predicate that returns true if the node identified by the given nodeId should be restarted
+     * @param abruptShutdown if true, kafka broker will be abruptly killed.
+     */
+    void restartBrokers(Predicate<Integer> nodeIdPredicate, boolean abruptShutdown);
```
[the remark this answers is not in the document]
Comment (k-wall) [reply api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java]: Done.

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 7]
```diff
@@ -354,6 +361,29 @@ public synchronized void close() throws Exception {
         }
     }
 
+    /**
+     * Workaround for <a href="https://issues.apache.org/jira/browse/KAFKA-14287">KAFKA-14287</a>.
+     * with kraft, if we don't shut down the controller last, we sometimes see a hang.
+     */
+    private void orderedShutdownServers(Map<Integer, Server> servers, boolean await) {
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:369]: Done.

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java (hunk 4 above)
Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java]: What do you think to this solution? @SamBarker @franvila

**Review by SamBarker — APPROVED** [review]
Thanks @k-wall I think this is close

**Review by franvila — APPROVED** [review]
LGTM Thanks @k-wall!

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java [hunk 8]
```diff
@@ -112,6 +127,56 @@ public void kafkaClusterZookeeperRemoveBroker() throws Exception {
...
+                .kraftMode(kraft)
+                .build())) {
+            cluster.start();
+            assertThat(cluster.getNumOfBrokers()).isEqualTo(brokersNum);
+            verifyRecordRoundTrip(brokersNum, cluster);
+
+            var nodes = describeClusterNodes(cluster);
+
+            var brokersExpectedDown = nodes.stream().filter(n -> integerPredicate.test(n.id())).toList();
+            var brokerInUnexpectedState = new AtomicReference<AssertionFailedError>();
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:154]: Hey @SamBarker ...  Thanks for the suggestion .  I'm happy to merge your alternative.

I did want to voice something first.  To be honest, I don't especially like the callback much (either your version or mine).  I particularly dislike the fact that the intrinsic lock is held during execution of the user's callback. That's a code smell. Then we've got all that guff about what you can and can't do inside the callback and the threading stuff.  `<shiver/>`

I keep toying with alternatives that avoid the need for it.

1. separate API for the `KafkaCluster#stopBrokers(Predicate<Integer>..)`, `KafkaCluster#startBrokers(...)` that break the restart action into two.
1. or some kind of fluent API variation `KafkaCluster#stopBrokers(Predicate<Integer>..).andRestart()`

What do you think?

[the remark this answers is not in the document]
Comment (SamBarker) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:154]: Hmm. I agree on the smell. Initially I did think `restart` was the API we wanted to offer but I think instead `stop` and `start`  as distinct calls is making more sense given the test we actually want to write. I do still sort of think a callback is useful to offer, so users can take action when the broker has actually stopped/started as opposed to polling for that status.

We could perhaps offer a restart without the call back which provides a no-op listener to each of stop and start... 

I'm happy to go with your thoughts on merging this and refactoring to stop & start or updating this PR to include that change.

[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java:154]: Thanks.  I'll refactor turn restartBrokers into separate `stopBrokers` and `startBrokers` API calls.   In both cases, control won't return to the caller until the stop or start has actually completed, so I don't think there is a need for a callback.

* I aim for idempotency - stopping a broker that is is stopped will have no effect, starting a broker that is already started likewise.
* When removing a broker, the check for reassignment target will consider only running brokers.

**Comment by k-wall** [comment]
@SamBarker the API is refactored, please take another look.  I'm having a test reliability issue which I am still investigating.

**Comment by k-wall** [comment]
> @SamBarker the API is refactored, please take another look. I'm having a test reliability issue which I am still investigating.

Grrr, the test reliability issue is another manifestation of [#17640 ](https://github.com/containers/podman/issues/17640).  The issue is that sometimes I'm seeing the first restart fail with the  spurious  `com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.NoHttpResponseException: localhost:2375 failed to respond`. However, the broker's container has got sufficiently far to create the ZK ephemeral node.  Testcontainer tries to restart again, but this time it fails because the ZK ephemeral node exists already.     How I'm hating podman right now.

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 9]
```diff
@@ -354,6 +370,29 @@ public synchronized void close() throws Exception {
         }
     }
 
+    /**
+     * Workaround for <a href="https://issues.apache.org/jira/browse/KAFKA-14287">KAFKA-14287</a>.
+     * with kraft, if we don't shut down the controller last, we sometimes see a hang.
+     */
+    private void roleOrderedShutdown(Map<Integer, Server> servers, boolean await) {
```
[the remark this answers is not in the document]
Comment (SamBarker) [reply impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:399]: I take that as yes to 1. but we don't see it due to the hard kill.

Do you see what I'm worrying about in 2?

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 10]
```diff
@@ -339,6 +467,11 @@ public synchronized int getNumOfBrokers() {
         return brokers.size();
     }
 
+    @Override
+    public synchronized Set<Integer> getStoppedBrokers() {
+        return new HashSet<>(this.stoppedBrokers);
```
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java:472]:
```suggestion
        return Set.copyOf(this.stoppedBrokers);
```

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 11]
```diff
@@ -354,11 +392,39 @@ public synchronized void close() throws Exception {
...
+    }
+
     @Override
     public synchronized int getNumOfBrokers() {
         return servers.size();
     }
 
+    @Override
+    public synchronized Set<Integer> getStoppedBrokers() {
+        return new HashSet<>(this.stoppedServers);
```
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:425]:
```suggestion
        return Set.copyOf(this.stoppedBrokers);
```

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 12]
```diff
@@ -354,11 +392,39 @@ public synchronized void close() throws Exception {
...
+     */
+    private void roleOrderedShutdown(Map<Integer, Server> servers, boolean await) {
+        var justBrokers = servers.entrySet().stream()
+                .filter(e -> !portsAllocator.hasRegisteredPort(Listener.CONTROLLER, e.getKey()))
+                .map(Map.Entry::getValue)
+                .toList();
+        justBrokers.forEach(Server::shutdown);
+        if (await) {
+            justBrokers.forEach(Server::awaitShutdown);
+        }
```
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:407]: I can't decide if we should do `stoppedServers.putAll(justBrokers)` here or if doing it after this method returns is fine... 

I think its moot.

**Review by SamBarker — APPROVED** [review]
For the record this passes reliably for me on Fedora.

File: api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java [hunk 13]
```diff
@@ -42,6 +43,29 @@ public interface KafkaCluster extends AutoCloseable {
...
+     * exactly once, when all the requested brokers are stopped. The restart of the brokers will not commence until
+     * this method returns.  Behaviour is not defined if the callable throws unchecked exceptions.  Furthermore, it
+     * must not attempt to mutate the cluster (that is, make calls to methods such as {@link KafkaCluster#addBroker()}).
+     * There is no guarantee made about the thread used to execute the callback.
+     *
+     * @param nodeIdPredicate  predicate that returns true if the node identified by the given nodeId should be restarted
+     * @param terminationStyle the style of termination used to shut down the broker(s).
+     * @param onBrokersStopped callback to invoked when all brokers are stopped, before restart commences. may be null.
+     */
+    void restartBrokers(Predicate<Integer> nodeIdPredicate, TerminationStyle terminationStyle, Runnable onBrokersStopped);
```
[the remark this answers is not in the document]
Comment (k-wall) [reply api/src/main/java/io/kroxylicious/testing/kafka/api/KafkaCluster.java]: We opted to remove the callback.

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:399]: 1. I expect so.
2. Yes, I do.  At a minimum there might be an opportunity for a pull-up type refactor to factor out the duplication around the server/broker maps.   Like you say, it might be better modelled differently.    I don't feel the motivation to make this change at the moment tbh.

File: impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java]: I did that.

**Comment by k-wall** [comment]
@SamBarker it has been a long road, but hopefully this is now close.

File: impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java [hunk 14]
```diff
@@ -206,8 +207,10 @@ private static KafkaFuture<Void> createTopic(Integer expectedBrokerCount, Admin
                     log.debug("Create topic future completed.");
                     if (throwable != null) {
                         log.warn("Failed to create topic: {} due to {}", CONSISTENCY_TEST, throwable.getMessage(), throwable);
-                        if (throwable instanceof RetriableException || throwable instanceof InvalidReplicationFactorException) {
-                            CompletableFuture.supplyAsync(() -> createTopic(expectedBrokerCount, admin));
+                        if (throwable instanceof RetriableException || throwable instanceof InvalidReplicationFactorException
+                                || (throwable instanceof TopicExistsException && throwable.getMessage().contains("is marked for deletion"))) {
+                            // Just retry after a short delay
+                            CompletableFuture.supplyAsync(() -> createTopic(expectedBrokerCount, admin), CompletableFuture.delayedExecutor(100, TimeUnit.MILLISECONDS));
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/common/Utils.java:213]: I've got a refactoring to fix this (existing issue), but I'll do it as a separate PR.

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java:373]: Re 2. As long as we are on the same page. I'll put trying to tease things about on my todo list but I want to try and make some progress on shared state first.

File: impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java [hunk 15]
```diff
@@ -320,21 +322,35 @@ public synchronized void removeBroker(int nodeId) throws IllegalArgumentExceptio
...
+        var kafkaServersToRestart = servers.entrySet().stream().filter(e -> nodeIdPredicate.test(e.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        roleOrderedShutdown(kafkaServersToRestart, true);
+
+        try {
+            if (onBrokersStopped != null) {
+                onBrokersStopped.run();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java]: we changed the API, so this comment is no longer relevant

**Comment by k-wall** [comment]
@SamBarker @franvila can I ask you both to re-review and test locally?
In the end I decided to use a named volume ([docker](https://docs.docker.com/storage/volumes/))([podman](https://docs.podman.io/en/latest/markdown/podman-volume-create.1.html) for the kafka log dir when running under test-containers.    This gives the semantics needed for the broker restart use-case without the added complication of actually 
mounting a directory from the host.

You'll see the wrinkle in `createNamedVolume` - the need to chown the volume in the Docker case.

It is working for me under both Docker Desktop and Podman on the Mac.  Looking forward to hearing how it fares on Linux desktops.

**Comment by franvila** [comment]
Tested and running fine on podman with linux

**Review by franvila — APPROVED** [review]

**Review by SamBarker — APPROVED** [review]
My issues appears to unrelated to these changes. 

Reproduced similar behaviour on main
```
❯ jstack -l 488731
2023-07-11 05:22:46
Full thread dump OpenJDK 64-Bit Server VM (20+36 mixed mode, sharing):

Threads class SMR info:
_java_thread_list=0x00007fc9d0001a40, length=13, elements={
0x00007fca2c029c90, 0x00007fca2c166090, 0x00007fca2c1675c0, 0x00007fca2c168f80,
0x00007fca2c16a410, 0x00007fca2c16b890, 0x00007fca2c16d330, 0x00007fca2c16e8d0,
0x00007fca2c17ccb0, 0x00007fca2c180cf0, 0x00007fca2c53b620, 0x00007fca2c54d610,
0x00007fc9d0000ff0
}

"main" #1 [488732] prio=5 os_prio=0 cpu=2607.61ms elapsed=88.72s tid=0x00007fca2c029c90 nid=488732 runnable  [0x00007fca31b19000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.SocketDispatcher.read0(java.base@20/Native Method)
	at sun.nio.ch.SocketDispatcher.read(java.base@20/SocketDispatcher.java:47)
	at sun.nio.ch.SocketChannelImpl.tryRead(java.base@20/SocketChannelImpl.java:1261)
	at sun.nio.ch.SocketChannelImpl.blockingRead(java.base@20/SocketChannelImpl.java:1336)
	at sun.nio.ch.SocketInputStream.read(java.base@20/SocketInputStream.java:70)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:149)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:280)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:241)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:53)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.impl.io.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:187)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.impl.io.HttpRequestExecutor.execute(HttpRequestExecutor.java:175)
	at com.github.dockerjava.zerodep.HijackingHttpRequestExecutor.execute(HijackingHttpRequestExecutor.java:50)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.impl.io.HttpRequestExecutor.execute(HttpRequestExecutor.java:218)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager$InternalConnectionEndpoint.execute(PoolingHttpClientConnectionManager.java:596)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.InternalExecRuntime.execute(InternalExecRuntime.java:215)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.MainClientExec.execute(MainClientExec.java:107)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ConnectExec.execute(ConnectExec.java:181)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ProtocolExec.execute(ProtocolExec.java:172)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.HttpRequestRetryExec.execute(HttpRequestRetryExec.java:93)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ContentCompressionExec.execute(ContentCompressionExec.java:128)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.RedirectExec.execute(RedirectExec.java:116)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.InternalHttpClient.doExecute(InternalHttpClient.java:178)
	at com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.CloseableHttpClient.execute(CloseableHttpClient.java:67)
	at com.github.dockerjava.zerodep.ApacheDockerHttpClientImpl.execute(ApacheDockerHttpClientImpl.java:191)
	at com.github.dockerjava.zerodep.ZerodepDockerHttpClient.execute(ZerodepDockerHttpClient.java:8)
	at org.testcontainers.dockerclient.HeadersAddingDockerHttpClient.execute(HeadersAddingDockerHttpClient.java:23)
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:228)
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.get(DefaultInvocationBuilder.java:202)
	at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.get(DefaultInvocationBuilder.java:74)
	at org.testcontainers.shaded.com.github.dockerjava.core.exec.InfoCmdExec.exec(InfoCmdExec.java:24)
	at org.testcontainers.shaded.com.github.dockerjava.core.exec.InfoCmdExec.exec(InfoCmdExec.java:14)
	at org.testcontainers.shaded.com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.tryOutStrategy(DockerClientProviderStrategy.java:291)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.lambda$getFirstValidStrategy$5(DockerClientProviderStrategy.java:265)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy$$Lambda$689/0x00000008012d8240.test(Unknown Source)
	at java.util.stream.ReferencePipeline$2$1.accept(java.base@20/ReferencePipeline.java:178)
	at java.util.stream.ReferencePipeline$2$1.accept(java.base@20/ReferencePipeline.java:179)
	at java.util.stream.ReferencePipeline$2$1.accept(java.base@20/ReferencePipeline.java:179)
	at java.util.ArrayList$ArrayListSpliterator.tryAdvance(java.base@20/ArrayList.java:1602)
	at java.util.stream.ReferencePipeline.forEachWithCancel(java.base@20/ReferencePipeline.java:129)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(java.base@20/AbstractPipeline.java:527)
	at java.util.stream.AbstractPipeline.copyInto(java.base@20/AbstractPipeline.java:513)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(java.base@20/AbstractPipeline.java:499)
	at java.util.stream.FindOps$FindOp.evaluateSequential(java.base@20/FindOps.java:150)
	at java.util.stream.AbstractPipeline.evaluate(java.base@20/AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(java.base@20/ReferencePipeline.java:647)
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:266)
	at org.testcontainers.DockerClientFactory.getOrInitializeStrategy(DockerClientFactory.java:150)
	- locked <0x0000000629001578> (a [Ljava.lang.Object;)
	at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:191)
	- locked <0x0000000629001578> (a [Ljava.lang.Object;)
	at org.testcontainers.DockerClientFactory$1.getDockerClient(DockerClientFactory.java:104)
	at com.github.dockerjava.api.DockerClientDelegate.authConfig(DockerClientDelegate.java:109)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:321)
	at io.kroxylicious.testing.kafka.testcontainers.TestcontainersKafkaCluster.start(TestcontainersKafkaCluster.java:237)
	```
	
Thanks for all the hard work @k-wall .

File: impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java [hunk 16]
```diff
@@ -399,12 +574,64 @@ private EndpointPair buildExposedEndpoint(Listener listener, int nodeId, int bin
...
+            // On Docker, use a container to chown the volume.
+            // This is a workaround for https://github.com/moby/moby/issues/45714
+            try (var c = new OneShotContainer()) {
+                c.withName("prepareKafkaVolume")
+                        .addGenericBind(new Bind(volumeName, new Volume(KAFKA_CONTAINER_MOUNT_POINT)))
+                        .withCommand("chown", "-R", KAFKA_CONTAINER_UID, KAFKA_CONTAINER_MOUNT_POINT)
+                        .withStartupCheckStrategy(new OneShotStartupCheckStrategy());
+                c.start();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaCluster.java]: The kafka image runs as uid 1001.  The issue that this is working around is the fact that named volume is owned by root, so 1001 will fail to write to it.  The OneShotContainer runs (as root) a chown to 1001 to resolve this problem. On Podman we don't need this step since it lets you specify a uid when you create the named volume.
